src/golang/cmd/server/middleware/usage/usage.go

```diff
@@ -0,0 +1,86 @@
+package usage
+
+import (
+	"bytes"
+	"encoding/json"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/denisbrodbeck/machineid"
+	"github.com/google/uuid"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	obfuscated string = "***"
+	delimiter  string = "/"
+	hashKey    string = "aqueduct"
+	// This is the Grafana Loki server address
+	logURL string = "http://34.25.83.71:9090/loki/api/v1/push"
+)
+
+type UsageStats struct {
+	ID      string `json:"id"`
+	Route   string `json:"route"`
+	Latency int64  `json:"latency"`
+}
+
+func reportUsage(startTime time.Time, r *http.Request) {
+	pathToken := strings.Split(r.URL.Path, delimiter)
+	for i, token := range pathToken {
+		if _, err := uuid.Parse(token); err == nil {
+			pathToken[i] = obfuscated
+		}
+	}
+
+	// This call generates a unique hash of the host device in a privacy-preserving fashion.
+	// Details can be found here: https://github.com/denisbrodbeck/machineid
+	machineID, err := machineid.ProtectedID(hashKey)
+	if err != nil {
+		log.Errorf("Failed to generate obfuscated device ID: %v", err)
+		return
+	}
+
+	usage := UsageStats{
+		ID:      machineID,
+		Route:   strings.Join(pathToken, delimiter),
+		Latency: time.Since(startTime).Milliseconds(),
+	}
+
+	log.Errorf("This request took %d ms, request URL path: %s, machine id: %s.", usage.Latency, usage.Route, usage.ID)
```

Comment:
Should this be log.Info()?

Reply:
This was actually meant to be a debug log. I'll remove it when merging.